### PR TITLE
Add support for adding dacpac references in sql database projects

### PR DESCRIPTION
**Pull request by kisantia**
This PR adds basic support for adding a reference to master or to another database's dacpac. Right now, the quickpick menu and input boxes are being used to get the information for adding the reference, but this can be changed to be dialog later. 

For adding a reference to master, the DSP in the sqlproj is used to determine which master.dacpac should be referenced. 
addMasterReference

Adding a reference to a dacpac requires the dacpac file location, where the database being referenced is (same database or different database on the same server), and the name of the database.
addDacpacReference

### CODE REVIEW

File: extensions/sql-database-projects/src/common/constants.ts [hunk 1]
```diff
@@ -91,7 +105,10 @@ export const Condition = 'Condition';
...
 export const Version = 'Version';
 export const PrivateAssets = 'PrivateAssets';
-
+export const ArtifactReference = 'ArtifactReference';
+export const SuppressMissingDependenciesErrors = 'SuppressMissingDependenciesErrors';
+export const DatabaseVariableLiteralValue = 'DatabaseVariableLiteralValue';
+export const DSP = 'DSP';
 
 // SqlProj File targets
 export const NetCoreTargets = '$(NETCoreTargetsPath)\\Microsoft.Data.Tools.Schema.SqlTasks.targets';
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/common/constants.ts:114]: I know this is not your change, just want to check: does this work for non-Windows platform?

File: extensions/sql-database-projects/src/models/project.ts [hunk 2]
```diff
@@ -277,7 +346,33 @@ export class ProjectEntry {
...
+	Folder,
+	DatabaseReference
+}
+
+export enum DatabaseReferenceLocation {
+	sameDatabase,
+	differentDatabaseSameServer
+}
+
+export enum SqlPlatforms {
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/models/project.ts:369]: is this name accurate? this seems like CompatLevels

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 3]
```diff
@@ -292,6 +292,133 @@ export class ProjectsController {
...
+	 * Adds a database reference to the project
+	 * @param treeNode a treeItem in a project's hierarchy, to be used to obtain a Project
+	 */
+	public async addDatabaseReference(treeNode: BaseProjectTreeItem): Promise<void>;
+	/**
+	 * Adds a database reference to the project
+	 * @param project Project to be built and deployed
+	 */
+	public async addDatabaseReference(project: Project): Promise<void>;
+	public async addDatabaseReference(context: Project | BaseProjectTreeItem): Promise<void> {
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/controllers/projectController.ts:305]: it seems to me you only need the third method signature, it covers the other 2 signatures.

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 4]
```diff
@@ -292,6 +292,133 @@ export class ProjectsController {
...
+			if (!databaseReferenceType) {
+				throw new Error(constants.databaseReferenceTypeRequired);
+			}
+
+			// if master is selected, we know which dacpac needs to be added
+			if (databaseReferenceType === constants.master) {
+				await project.addMasterDatabaseReference();
+			} else {
+				// get other information needed to add a reference to the dacpac
+				let dacpacFileLocation = await this.getDacpacFileLocation();
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/controllers/projectController.ts:321]: do you also need to check the existence of the file?

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 5]
```diff
@@ -292,6 +292,133 @@ export class ProjectsController {
...
+	private async getDacpacFileLocation(): Promise<Uri | undefined> {
+		let fileUris = await this.apiWrapper.showOpenDialog(
+			{
+				canSelectFiles: true,
+				canSelectFolders: false,
+				canSelectMany: false,
+				defaultUri: this.apiWrapper.workspaceFolders() ? (this.apiWrapper.workspaceFolders() as WorkspaceFolder[])[0].uri : undefined,
+				openLabel: constants.selectString,
+				filters: {
+					'dacpac Files': ['dacpac'],
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/controllers/projectController.ts:376]: localize 'dacpac Files'?

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 6]
```diff
@@ -292,6 +292,133 @@ export class ProjectsController {
...
+	}
+
+	private async getDatabaseName(dacpac: Uri): Promise<string | undefined> {
+		const dacpacName = path.parse(dacpac.toString()).name;
+		let databaseName = await this.apiWrapper.showInputBox({
+			prompt: constants.databaseReferenceDatabaseName,
+			value: `${dacpacName}`
+		});
+
+		databaseName = databaseName?.trim();
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/controllers/projectController.ts:418]: please validate the database name

File: extensions/sql-database-projects/src/models/project.ts [hunk 7]
```diff
@@ -141,6 +141,53 @@ export class Project {
...
 	}
 
+	/**
+	 * Set the compat level of the project
+	 * Just used in tests right now, but can be used later if this functionality is added to the UI
+	 * @param compatLevel compat level of project
+	 */
+	public changeDSP(compatLevel: string): void {
+		const newDSP = `${constants.MicrosoftDatatoolsSchemaSqlSql}${compatLevel}${constants.databaseSchemaProvider}`;
+		this.projFileXmlDoc.getElementsByTagName(constants.DSP)[0].childNodes[0].nodeValue = newDSP;
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/models/project.ts:151]: i am not familiar with the xml schema, is it ok to always use the first DSP and first child node of DSP??

File: extensions/sql-database-projects/src/models/project.ts [hunk 8]
```diff
@@ -141,6 +141,53 @@ export class Project {
...
+	}
+
+	/**
+	 * Adds reference to the appropriate master dacpac to the project
+	 */
+	public async addMasterDatabaseReference(): Promise<void> {
+		const uri = this.getMasterDacpac();
+		this.addDatabaseReference(uri, DatabaseReferenceLocation.differentDatabaseSameServer, 'master');
```
Comment (alanrenmsft) [inline extensions/sql-database-projects/src/models/project.ts:159]: please use the master constant

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 9]
```diff
@@ -292,6 +292,133 @@ export class ProjectsController {
...
+
+			// if master is selected, we know which dacpac needs to be added
+			if (databaseReferenceType === constants.master) {
+				await project.addMasterDatabaseReference();
+			} else {
+				// get other information needed to add a reference to the dacpac
+				let dacpacFileLocation = await this.getDacpacFileLocation();
+				if (!dacpacFileLocation) {
+					throw new Error(constants.dacpacFileLocationRequired);
+				}
```
[the remark this answers is not in the document]
Comment (kisantia) [reply extensions/sql-database-projects/src/controllers/projectController.ts:324]: I was following how it was done in importNewDatabaseProject(), but I can move the error throwing into getDacpacFileLocation() if that's more clear

File: extensions/sql-database-projects/src/test/project.test.ts [hunk 10]
```diff
@@ -81,4 +82,30 @@ describe('Project: sqlproj content operations', function (): void {
...
 		await testUtils.shouldThrowSpecificError(async () => await project.addToProject(list), `ENOENT: no such file or directory, stat \'${nonexistentFile}\'`);
 	});
+
+	it('Should choose correct master dacpac', async function(): Promise<void> {
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.newProjectFileBaseline);
+		const project = new Project(projFilePath);
+		await project.readProjFile();
+
+		let uri = project.getMasterDacpac();
+		should.equal(uri.fsPath, '/$(NETCoreTargetsPath)/SystemDacpacs/130/master.dacpac');
```
[the remark this answers is not in the document]
Comment (kisantia) [reply extensions/sql-database-projects/src/test/project.test.ts]: nope it failed. Thanks for catching this

File: extensions/sql-database-projects/src/controllers/projectController.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (kisantia) [reply extensions/sql-database-projects/src/controllers/projectController.ts]: I moved the error throwing into the get methods

File: extensions/sql-database-projects/src/models/project.ts (hunk 7 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/models/project.ts:151]: added validation to make sure there's only one DSP element tag and it has one child

File: extensions/sql-database-projects/src/models/project.ts (hunk 2 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/models/project.ts]: changed this to TargetPlatform to match SSDT

File: extensions/sql-database-projects/src/controllers/projectController.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Benjin) [reply extensions/sql-database-projects/src/controllers/projectController.ts]: Thanks!  I think Sakshi's doing a pass on that front over in `importNewDatabaseProject()`, too.

File: extensions/sql-database-projects/src/test/project.test.ts [hunk 11]
```diff
@@ -82,6 +86,58 @@ describe('Project: sqlproj content operations', function (): void {
...
+			should.equal(uri.fsPath, '/$(NETCoreTargetsPath)/SystemDacpacs/130/master.dacpac');
+
+			project.changeDSP(TargetPlatform.Sql150.toString());
+			uri = project.getMasterDacpac();
+			should.equal(uri.fsPath, '/$(NETCoreTargetsPath)/SystemDacpacs/150/master.dacpac');
+
+			project.changeDSP(TargetPlatform.SqlAzureV12.toString());
+			uri = project.getMasterDacpac();
+			should.equal(uri.fsPath, '/$(NETCoreTargetsPath)/SystemDacpacs/AzureV12/master.dacpac');
+		}
```
Comment (SakshiS-harma) [inline extensions/sql-database-projects/src/test/project.test.ts:130]: Can we instead use path.join to get platform independent path location (and then verifying it again uri.fsPath)?

**Pull request merged**